dipy/tracking/vox2track.pyx

```diff
@@ -162,6 +161,176 @@ def streamline_mapping(streamlines, voxel_size=None, affine=None,
+    cdef cnp.double_t[:] direction = np.zeros(3, dtype=np.double)
+    cdef cnp.double_t[:] current_edge = np.zeros(3, dtype=np.double)
+
+    cdef int nb_streamlines = len(streamlines)
+    cdef cnp.uint8_t[:] in_mask = np.zeros(nb_streamlines, dtype=np.uint8)
+    cdef int streamline_idx
+
+    for streamline_idx in range(nb_streamlines):
+        # Can't call _to_voxel_coordinates because it casts to int
+        voxel_indices = np.dot(streamlines[streamline_idx], lin_T) + offset
+
+        in_mask[streamline_idx] = _streamline_in_mask(
+            voxel_indices, mask, current_pt, next_pt, direction, current_edge)
+
+    return np.asarray(in_mask)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef int _streamline_in_mask(cnp.double_t[:,:] streamline,
+                             cnp.uint8_t[:,:,:] mask,
+                             cnp.double_t[:] current_pt,
+                             cnp.double_t[:] next_pt,
+                             cnp.double_t[:] direction,
+                             cnp.double_t[:] current_edge) nogil:
+    """
+    Check if a single streamline is passing through a mask. This ia an utility
+    function to make streamlines_in_mask() more readable.
+    """
+    cdef cnp.double_t direction_norm, remaining_distance
+    cdef cnp.double_t length_ratio, half_ratio
+    cdef int point_idx, dim_idx
+    cdef int x, y, z
+
+    if streamline.shape[0] <= 1:
+        return 2  # Single-point streamline
+
+    # This loop is time-critical
+    # Changed to -1 because we get the next point in the loop
+    for point_idx in range(streamline.shape[0] - 1):
+        # Assign current and next point, find vector between both,
+        # and use the current point as nearest edge for testing.
+        for dim_idx in range(3):
+            current_pt[dim_idx] = streamline[point_idx, dim_idx]
+            next_pt[dim_idx] = streamline[point_idx + 1, dim_idx]
+            direction[dim_idx] = next_pt[dim_idx] - current_pt[dim_idx]
+            current_edge[dim_idx] = current_pt[dim_idx]
+
+        # Compute norm
+        direction_norm = norm(direction[0], direction[1], direction[2])
+
+        # Set the "dist" var to compute remaining length of vector
+        # to process
+        remaining_distance = direction_norm
+
+        # Check if it's already a real edge. If not, find the closest edge.
+        if floor(current_edge[0]) != current_edge[0] and \
+           floor(current_edge[1]) != current_edge[1] and \
+           floor(current_edge[2]) != current_edge[2]:
+            # All coordinates are not "integers", and therefore, not on the
+            # edge. Fetch the closest edge.
+            c_get_closest_edge(current_pt[0], current_pt[1], current_pt[2],
+                               direction[0], direction[1], direction[2],
+                               current_edge)
+
+        # TODO Could condition be optimized?
```

Comment:
I tested and, as far as I can tell, it wasn't faster and it forced me to 1) still use a `break` or 2) put everything in an `else`, which I don't want to do because it's already indented enough and we have a 80c maximum.